### PR TITLE
kubeapiserver: rename '--experimental-encryption-provider-config' to …

### PR DESCRIPTION
**Pull request by marrrvin**
…'--encryption-provider-config'.

This change renames the '--experimental-encryption-provider-config' flag to '--encryption-provider-config'. The old flag is accepted but generates a warning.

In 1.12, we will drop support for '--experimental-encryption-provider-config' entirely.

**What this PR does / why we need it**:

Encryption at rest with static configs has been promoted to beta.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Updates #61420

**Special notes for your reviewer**:

**Release note**:

```release-note
Encryption at rest with static configs has been promoted to beta.
API server flag `--experimental-encryption-provider-config` has been renamed to `--encryption-provider-config`. The old flag is accepted with a warning but will be removed in 1.12.
```

### CODE REVIEW

**Comment by marrrvin** [comment]
@php-coder FYI

**Comment by php-coder** [comment]
@marrrvin Thank you for your contribution!

Could you update PR title to make it shorter?

Also in the description, please, replace `Fixes #` by `Fixes #61420` (in this case when PR will get merged, the related issue will be closed automatically).

And this issue deserves to be mentioned in the changelog, hence release note is needed. I'd suggest to something like:
```
Encryption at rest with static configs has been promoted to beta.
API server flag `--experimental-encryption-provider-config` has been renamed to `--encryption-provider-config`. The old flag is accepted with a warning but will be removed in 1.12
```

@ericchiang @liggitt PTAL. Do we need release note with required action in this case?

/sig auth
/ok-to-test

**Comment by liggitt** [comment]
before this can be merged, we have to fix the way the encryption config is parsed. currently it uses a non-standard parsing method that pays no attention to the version of the config:

https://github.com/kubernetes/kubernetes/blob/dce1b881284a103909f5cfa969ff56e5e0565362/staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go#L67-L79

the types must change to a versioned config (we can promote straight to v1beta1 if we don't require any structural changes), and parsing must require well-formed config files before this is promoted out of experimental mode.

**Comment by marrrvin** [comment]
@liggitt Thanks. Could you please put me an example of standard parsing method?

**Comment by php-coder** [comment]
> Could you please put me an example of standard parsing method?

Looks like you could follow this example:

https://github.com/kubernetes/kubernetes/blob/dce1b881284a103909f5cfa969ff56e5e0565362/cmd/kube-proxy/app/server.go#L292-L314

@liggitt Please, confirm.

**Comment by liggitt** [comment]
yes, roughly that.

1. fix type definition to actually be a runtime object, inline TypeMeta, etc
2. define v1beta1 API types, generate conversions/defaults
3. define scheme/codecs that have the external (v1beta1) and internal (existing) types registered
4. use that scheme/codec to load the config (handles decoding the v1beta1 version, conversion to internal version)

**Comment by php-coder** [comment]
@liggitt Thanks, Jordan! I've extracted this task to a separate issue: #61599

**Review by dims — APPROVED** [review]
Looks like changes requested by @liggitt was done in another PR

**Comment by php-coder** [comment]
> Looks like changes requested by @liggitt was done in another PR

Not yet, unfortunately (as now that PR requires another PR :) )

**Review by php-coder — CHANGES_REQUESTED** [review]
Besides rebasing against the latest master, we need to also update another place that was introduced recently:
```console
$ git grep 'experimental-encryption-provider-config'
...
cluster/gce/gci/apiserver_manifest_test.go:     return fmt.Sprintf("--experimental-encryption-provider-config=%s", path)
cluster/gce/gci/apiserver_manifest_test.go:// EncryptionProviderConfig file of type KMS and inject experimental-encryption-provider-config startup flag.
```

**Comment by marrrvin** [comment]
@php-coder ok. I'll try to do it on holidays.

**Comment by dims** [comment]
/milestone v1.12

Looks like this needs to be in 1.12 please drop the milestone if i am reading this wrong

**Comment by dims** [comment]
@marrrvin can you please rebase?

**Comment by liggitt** [comment]
/hold

while this is definitely desired, we shouldn't drop the experimental prefix until the config driving this graduates from alpha (in progress in https://github.com/kubernetes/kubernetes/pull/67383)

**Comment by shubheksha** [comment]
ping @marrrvin, @liggitt - is this on track for 1.13 since there hasn't been any update?

**Comment by dims** [comment]
@marrrvin  this needs rebase ... code freeze is tomorrow! ( https://github.com/kubernetes/sig-release/tree/master/releases/release-1.13 )

**Comment by stlaz** [comment]
@dims should we just do the rebase ourselves? I've got it ready in https://github.com/stlaz/kubernetes/tree/enc_config_opt but won't be able to push it in this PR.
The hold could also be dropped since the alpha->stable PR was pushed already.

**Comment by dims** [comment]
@stlaz please open a fresh PR and we can discuss there.

**Comment by stlaz** [comment]
@dims Done in https://github.com/kubernetes/kubernetes/pull/71206

**Comment by stlaz** [comment]
This PR can now be closed as the above one was merged.

**Pull request closed without merging**
